sgkit/stats/regenie.py

```diff
@@ -0,0 +1,778 @@
+from typing import Any, Dict, Hashable, Optional, Sequence, Tuple, Union
+
+import dask.array as da
+import numpy as np
+import xarray as xr
+from dask.array import Array
+from numpy import ndarray
+from xarray import Dataset
+
+from ..typing import ArrayLike
+from ..utils import split_array_chunks
+from .utils import (
+    assert_array_shape,
+    assert_block_shape,
+    assert_chunk_shape,
+    concat_2d,
+    r2_score,
+)
+
+
+def index_array_blocks(
+    x: Union[ArrayLike, Sequence[int]], size: int
+) -> Tuple[ndarray, ndarray]:
+    """Generate indexes for blocks that partition an array within groups.
+
+    Given an array with monotonic increasing group assignments (as integers),
+    this function will generate the indexes of blocks within those groups that
+    are of at most `size` elements.
+
+    Parameters
+    ----------
+    x : Union[ArrayLike, Sequence[int]]
+        Vector of group assignments, must be monotonic increasing.
+        Resulting blocks will never cross these group assignments
+        and the resulting `index` and `sizes` values constitute
+        covering slices for any array of the same size as `x`.
+    size : int
+        Maximum block size.
+
+    Examples
+    --------
+    >>> from sgkit.stats.regenie import index_array_blocks
+    >>> index_array_blocks([0, 0, 0], 2)
+    (array([0, 2]), array([2, 1]))
+    >>> index_array_blocks([0, 0, 1, 1, 1], 2)
+    (array([0, 2, 4]), array([2, 2, 1]))
+
+    Returns
+    -------
+    index : ndarray
+        Array of indexes for each block start
+    sizes : ndarray
+        Size of block such that `x[index[0]:(index[0] + sizes[0])]` contains
+        every element in block 0
+
+    Raises
+    ------
+    ValueError
+        If `x` is not 1D.
+    ValueError
+        If `size` is <= 0.
+    ValueError
+        If `x` does not contain integers.
+    ValueError
+        If `x` is not monotonic increasing.
+    """
+    x = np.asarray(x)
+    if x.size == 0:
+        return np.empty(0, dtype=int), np.empty(0, dtype=int)
+    if x.ndim != 1:
+        raise ValueError(f"Array shape {x.shape} is not 1D")
+    if size <= 0:
+        raise ValueError(f"Block size {size} must be > 0")
+    if not np.issubdtype(x.dtype, np.integer):
+        raise ValueError("Array to partition must contain integers")
+    if np.any(np.diff(x) < 0):
+        raise ValueError("Array to partition must be monotonic increasing")
+    breaks = np.argwhere(np.diff(x, prepend=x[0]))[:, 0]
+    breaks = np.concatenate(([0], breaks, [x.size]))
+    index = np.concatenate(
+        [np.arange(breaks[i], breaks[i + 1], size) for i in range(breaks.size - 1)]
+    )
+    sizes = np.diff(index, append=x.size)
+    assert index.size == sizes.size
+    return index, sizes
+
+
+def index_block_sizes(
+    sizes: Union[ArrayLike, Sequence[int]]
+) -> Tuple[ndarray, ndarray]:
```

Comment:
Should this be `Tuple[ArrayLike, ArrayLike]`?

Reply:
That was intentional because the assumption is that block metadata always fits in memory and I think return types should be more general if there's ever really a chance they'd be any other kind of array, but there's no scenario I can think of where that would be true in this case.